### PR TITLE
Feature: configuration files mounting

### PR DESCRIPTION
**Pull request by itamar-marom**
As discussed in Slack, I Added the ability to add configuration files before the runner initializes terraform.
#322 

API:
```yaml
spec:
  fileMappings:
  - secretRef:
      name: git-config      # must be local to the current namespace. we don't allow violations across the namespace boundary.
      key: git-config-content-key # key inside the secret to refer.
    location: home          # can be "home" or "workspace"
    filename: .git/config   # must be relative to "location" and we'll have to block the path traversal exploit.
  - secretRef:
      name: override-tf
      key: ovrdtf-content-key
    location: workspace     # workspace is the directory where we initialized Terraform
    filename: ./override.tf.json
```

### CODE REVIEW

**Comment by itamar-marom** [comment]
An example terraform:
```yaml
apiVersion: infra.contrib.fluxcd.io/v1alpha1
kind: Terraform
metadata:
  name: test
  namespace: tf-system
spec:
  approvePlan: "auto"
  destroyResourcesOnDeletion: true
  refreshBeforeApply: true
  interval: 30s
  path: YOUR PATH
  runnerPodTemplate:
    spec:
      fileMappings:
        - location: "home"
          path: ".gitconfig"
          secretRef:
            name: gitconfig
            key: mygitconfig
        - location: "home"
          path: ".git-credentials"
          secretRef:
            name: git-credetials
            key: mygitdredentials
  sourceRef:
    kind: GitRepository
    name: local-environment
    namespace: tf-system
```

File: runner/server.go [hunk 1]
```diff
@@ -216,6 +219,40 @@ func (r *TerraformRunnerServer) SetEnv(ctx context.Context, req *SetEnvRequest)
 	return &SetEnvReply{Message: "ok"}, nil
 }
 
+func (r *TerraformRunnerServer) CreateFileMappings(ctx context.Context, req *CreateFileMappingsRequest) (*CreateFileMappingsReply, error) {
+	log := ctrl.LoggerFrom(ctx).WithName(loggerName)
+	log.Info("creating file mappings")
+
+	for _, fileMapping := range req.FileMappings {
+		logMessage := fmt.Sprintf("fileMapping: content: %s, location: %s, path: %s", fileMapping.Content, fileMapping.Location, fileMapping.Path)
```
Comment (bivas) [inline runner/server.go:227]: Looks like a clear case for structured logging instead of all this string formatting

File: runner/server.go [hunk 2]
```diff
@@ -216,6 +219,40 @@ func (r *TerraformRunnerServer) SetEnv(ctx context.Context, req *SetEnvRequest)
...
+		log.Info(logMessage)
+		var fileFullPath string
+		switch fileMapping.Location {
+		case runnerFileMappingLocationHome:
+			fileFullPath = filepath.Join(HomePath, fileMapping.Path)
+		case runnerFileMappingLocationWorkspace:
+			fileFullPath = filepath.Join(req.WorkingDir, fileMapping.Path)
+		}
+		logMessage = fmt.Sprintf("fullPath: %s", fileFullPath)
+		log.Info(logMessage)
```
Comment (bivas) [inline runner/server.go:237]: Not really saving much with formatter in previous line

File: controllers/terraform_controller.go [hunk 3]
```diff
@@ -915,6 +945,30 @@ terraform {
 	return terraform, tfInstance, tmpDir, nil
 }
 
+func (r *TerraformReconciler) createRunnerFileMapping(ctx context.Context, terraform infrav1.Terraform) ([]*runner.FileMapping, error) {
+	var runnerFileMappingList []*runner.FileMapping
+
+	for _, fileMapping := range terraform.Spec.RunnerPodTemplate.Spec.FileMappings {
+		var runnerFileMapping runner.FileMapping
```
Comment (bivas) [inline controllers/terraform_controller.go:952]: Why do you need this fwd declaration? 
You can init the struct in-place when appending

File: controllers/terraform_controller.go [hunk 4]
```diff
@@ -822,6 +824,34 @@ terraform {
 		), tfInstance, tmpDir, err
 	}
 
+	if len(terraform.Spec.RunnerPodTemplate.Spec.FileMappings) > 0 {
+		log.Info("generate runner mapping files")
+		runnerFileMappingList, err := r.createRunnerFileMapping(ctx, terraform)
+		if err != nil {
+			err = fmt.Errorf("error creating runner file mappings: %s", err)
```
Comment (bivas) [inline controllers/terraform_controller.go:831]: Should use `%w` when appending errors

File: controllers/terraform_controller.go [hunk 5]
```diff
@@ -822,6 +824,34 @@ terraform {
...
+				err.Error(),
+			), tfInstance, tmpDir, err
+		}
+
+		log.Info("create mapping files")
+		if _, err := runnerClient.CreateFileMappings(ctx, &runner.CreateFileMappingsRequest{
+			WorkingDir:   workingDir,
+			FileMappings: runnerFileMappingList,
+		}); err != nil {
+			err = fmt.Errorf("error creating file mappings for Terraform: %s", err)
```
Comment (bivas) [inline controllers/terraform_controller.go:845]: Same as above

File: runner/runner.proto [hunk 6]
```diff
@@ -59,6 +60,21 @@ message SetEnvReply {
   string message = 1;
 }
 
+message fileMapping {
+  string content = 1;
```
Comment (bivas) [inline runner/runner.proto:64]: Shouldn't this be `bytes`?

File: config/crd/bases/infra.contrib.fluxcd.io_terraforms.yaml [hunk 7]
```diff
@@ -1,10 +1,9 @@
-
 ---
 apiVersion: apiextensions.k8s.io/v1
 kind: CustomResourceDefinition
 metadata:
   annotations:
-    controller-gen.kubebuilder.io/version: v0.7.0
+    controller-gen.kubebuilder.io/version: v0.9.0
```
Comment (chanwit) [inline config/crd/bases/infra.contrib.fluxcd.io_terraforms.yaml:6]: could you downgrade `kubebuilder` to v0.7.x and re-run `make generate` and `make manifests` please?

**Review by chanwit — CHANGES_REQUESTED** [review]

File: config/crd/bases/infra.contrib.fluxcd.io_terraforms.yaml (hunk 7 above)
Comment (itamar-marom) [reply config/crd/bases/infra.contrib.fluxcd.io_terraforms.yaml:6]: sure

**Review by chanwit — APPROVED** [review]
LGTM
Thank you @itamar-marom :1st_place_medal:

**Pull request merged**